File: Java/t23.java
import java.util.ArrayList;
import java.util.Collections; 

/**
 * Definition for singly-linked list.
 */
class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }
}

/**
 * 将所有链表元素提取出来放在数组排序之后再逐个添加至新链表
*/
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {

        // 输入链表数组长度为1时直接返回该链表
        if(lists.length == 1){
            return lists[0];
        }

        // 取出所有元素放在数组中
        ArrayList<Integer> node_list = new ArrayList<Integer>();
        for(int i=0;i<lists.length;i++){
            while(lists[i] != null){
                node_list.add(lists[i].val);
                lists[i] = lists[i].next;
            }
        }
        
        // 排序
        Collections.sort(node_list);

        ListNode new_node = new ListNode(0);
        ListNode resut;
        resut = new_node;

        // 将排序后的数组元素逐个添加至新链表
        for(Integer i:node_list){
            new_node.next = new ListNode(i);
            new_node = new_node.next;
        }
        return resut.next;
    }

    /**
     * 打印显示链表的元素，用于结果验证
    */
    public void showListNode(ListNode node){
        while(node != null){
            System.out.print(node.val+"\t");
            node = node.next;
        }
    }
}


public class t23 {

    public static void main(String[] args){

        // 创建三个测试链表添加至链表数组
        ListNode node1 = new ListNode(1);
        node1.next = new ListNode(4);
        node1.next.next = new ListNode(5);

        ListNode node2 = new ListNode(1);
        node2.next = new ListNode(3);
        node2.next.next = new ListNode(4);

        ListNode node3 = new ListNode(2);
        node3.next = new ListNode(6);

        ListNode[] lists = {node1, node2, node3};

        Solution result = new Solution();
        ListNode res = result.mergeKLists(lists);;

        result.showListNode(res);
    }
    
}

File: Python/t23.py
# -*- conding:utf-8 -*-
#Author:lyc
import os, sys

# 题目：
# 合并 k 个排序链表，返回合并后的排序链表。请分析和描述算法的复杂度。
#
# 示例:
# 输入:
# [
#   1->4->5,
#   1->3->4,
#   2->6
# ]
# 输出: 1->1->2->3->4->4->5->6



# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def mergeKLists(self, lists: [ListNode]) -> ListNode:
        """
        将所有元素取出来放在列表中排序再逐个添加至新链表
        :param lists:
        :return:
        """

        # 当传入的链表list为空或者只有一个时直接返回
        if len(lists) == 0:return []
        if len(lists) == 1:return lists[0]

        # 定义最终返回的链表，带有保护节点
        merge_node = ListNode(0)
        result = merge_node

        # 将所有链表的元素取出放入列表
        node_list = []
        for node in lists:
            while node:
                node_list.append(node.val)
                node = node.next

        # 排序
        node_list.sort()

        # 将列表中每一个元素添加至新链表
        while node_list:
            merge_node.next = ListNode(node_list.pop(0))
            merge_node = merge_node.next

        return result.next


    def showNode(self, node : ListNode) -> list:
        """
        show all value of ListNode
        :param node:
        :return:
        """
        nodeValue = []
        head = node
        while head:
            nodeValue.append(head.val)
            head = head.next
        return nodeValue


class Solution:
    def mergeKLists(self, lists: [ListNode]) -> ListNode:
        """
        官网解法，使用基于堆的优先级队列
        :param lists:
        :return:
        """
        import heapq
        dummy = ListNode(0)
        p = dummy
        head = []
        for i in range(len(lists)):
            if lists[i] :
                heapq.heappush(head, (lists[i].val, i))
                lists[i] = lists[i].next
        while head:
            val, idx = heapq.heappop(head)
            p.next = ListNode(val)
            p = p.next
            if lists[idx]:
                heapq.heappush(head, (lists[idx].val, idx))
                lists[idx] = lists[idx].next
        return dummy.next

    def showNode(self, node : ListNode) -> list:
        """
        show all value of ListNode
        :param node:
        :return:
        """
        nodeValue = []
        head = node
        while head:
            nodeValue.append(head.val)
            head = head.next
        return nodeValue


# 创建三个测试链表
node1 = ListNode(1)
node1.next = ListNode(4)
node1.next.next = ListNode(5)

node2 = ListNode(1)
node2.next = ListNode(3)
node2.next.next = ListNode(4)

node3 = ListNode(2)
node3.next = ListNode(6)

lists = [node1, node2, node3]

result = Solution()
new_node = result.mergeKLists(lists)
print(result.showNode(new_node))












